monai/networks/nets/dints.py

```diff
@@ -0,0 +1,556 @@
+        result = 0
+        for _ in pos:
+            result += self._ops[_.item()](x) * ops[_.item()] * weight[_.item()]
+        return result
+
+
+class Cell(nn.Module):
+    """
+    The basic class for cell operation
+    Args:
+        c_prev: input channel number
+        C: output channel number
+        rate: resolution change rate. -1 for 2x downsample, 1 for 2x upsample
+              0 for no change of resolution
+        code_c: cell operation code
+    """
+
+    def __init__(self, c_prev, c, rate: int, code_c: bool = None):
+        super().__init__()
+        self.c_out = c
+        if rate == -1:  # downsample
+            self.preprocess = _FactorizedReduceBlockWithMemory(c_prev, c)
+        elif rate == 1:  # upsample
+            self.preprocess = _FactorizedIncreaseBlockWithMemory(c_prev, c)
+        else:
+            if c_prev == c:
+                self.preprocess = _IdentityWithMemory()
+            else:
+                self.preprocess = _ReLUConvBNBlockWithMemory(c_prev, c, 1, 0)
+        self.op = MixedOp(c, code_c)
+
+    def forward(self, s, ops, weight):
+        s = self.preprocess(s)
+        s = self.op(s, ops, weight)
+        return s
+
+
+class DiNTS(nn.Module):
+    def __init__(
+        self,
+        in_channels: int,
+        num_classes: int,
+        cell=Cell,
+        cell_ops: int = 5,
+        channel_mul: float = 1.0,
+        num_blocks: int = 6,
+        num_depths: int = 3,
+        use_stem: bool = False,
+        code: list = None,
+    ):
+        """
+        Initialize NAS network search space
+
+        Args:
+            in_channels: input image channel
+            num_classes: number of segmentation classes
+            num_blocks: number of blocks (depth in the horizontal direction)
+            num_depths: number of image resolutions: 1, 1/2, 1/4 ... in each dimention, each resolution feature
+                is a node at each block
+            cell: operatoin of each node
+            cell_ops: cell operation numbers
+            channel_mul: adjust intermediate channel number, default 1.
+            code: [node_a, code_a, code_c] decoded using self.decode(). Remove unused cells in retraining
+
+        Predefined variables:
+            filter_nums: default init 64. Double channel number after downsample
+            topology related varaibles from gen_mtx():
+                trans_mtx: feasible path activation given node activation key
+                code2in: path activation to its incoming node index
+                code2ops: path activation to operations of upsample 1, keep 0, downsample -1
+                code2out: path activation to its output node index
+                node_act_list: all node activation codes [2^num_depths-1, res_num]
+                node_act_dict: node activation code to its index
+                tidx: index used to convert path activation matrix (depth,depth) in trans_mtx to path activation
+                    code (1,3*depth-2)
```

Comment:
might be useful to include some examples like https://github.com/Project-MONAI/MONAI/blob/33b6d61fdd8a15247e63bf7db16ac32f98f83566/monai/networks/nets/unetr.py#L65-L66